File: flutter_application_1/lib/services/reservation_service.dart
import 'package:dio/dio.dart';
import 'package:shared_preferences/shared_preferences.dart';
import '../models/reservation_model.dart';
import 'utils/dio_client.dart';

class ReservationService {
  Future<Map<String, dynamic>> reserver(ReservationModel reservation) async {
    final prefs = await SharedPreferences.getInstance();
    final token = prefs.getString('access_token') ?? '';
    final dio = DioClient.create(token: token);

    try {
      final response = await dio.post(
        '/api/reservations',
        data: reservation.toJson(),
      );
      return response.data;
    } on DioException catch (e) {
      print('Erreur lors de la réservation: ${e.message}');
      throw Exception('Erreur lors de la réservation');
    }
  }

  Future<void> updateReservation(int reservationId, int camionId, bool isIgnore) async {
    final prefs = await SharedPreferences.getInstance();
    final token = prefs.getString('access_token') ?? '';
    final dio = DioClient.create(token: token);

    try {
      await dio.put(
        '/api/reservations/$reservationId/camion/$camionId/$isIgnore',
      );
    } on DioException catch (e) {
      print('Erreur lors de la mise à jour de la réservation: ${e.message}');
      throw Exception(e.response?.data ?? 'Erreur lors de la mise à jour de la réservation');
    }
  }

  // Récupérer toutes les réservations de l'utilisateur
  Future<List<Map<String, dynamic>>> getUserReservations() async {
    final prefs = await SharedPreferences.getInstance();
    final token = prefs.getString('access_token') ?? '';
    final dio = DioClient.create(token: token);

    try {
      final response = await dio.get('/api/reservations');
      return List<Map<String, dynamic>>.from(response.data);
    } on DioException catch (e) {
      print('Erreur lors de la récupération des réservations: ${e.message}');
      throw Exception('Erreur lors de la récupération des réservations');
    }
  }

  // Récupérer les réservations par statut
  Future<List<Map<String, dynamic>>> getUserReservationsByStatus(String status) async {
    final prefs = await SharedPreferences.getInstance();
    final token = prefs.getString('access_token') ?? '';
    final dio = DioClient.create(token: token);

    try {
      final response = await dio.get('/api/reservations/status/$status');
      return List<Map<String, dynamic>>.from(response.data);
    } on DioException catch (e) {
      print('Erreur lors de la récupération des réservations par statut: ${e.message}');
      throw Exception('Erreur lors de la récupération des réservations par statut');
    }
  }
  
  // Récupérer les statistiques du transporteur
  Future<Map<String, dynamic>> getTransporteurStatistics() async {
    final prefs = await SharedPreferences.getInstance();
    final token = prefs.getString('access_token') ?? '';
    final dio = DioClient.create(token: token);

    try {
      final response = await dio.get('/api/reservations/my/statistics');
      return Map<String, dynamic>.from(response.data);
    } on DioException catch (e) {
      print('Erreur lors de la récupération des statistiques: ${e.message}');
      throw Exception('Erreur lors de la récupération des statistiques');
    }
  }

  // Récupérer les récapitulatifs de réservations
  Future<List<Map<String, dynamic>>> getReservationRecapitulatif() async {
    final prefs = await SharedPreferences.getInstance();
    final token = prefs.getString('access_token') ?? '';
    final dio = DioClient.create(token: token);

    try {
      final response = await dio.get('/api/reservations/recapitulatif');
      return List<Map<String, dynamic>>.from(response.data);
    } on DioException catch (e) {
      print('Erreur lors de la récupération des récapitulatifs: ${e.message}');
      throw Exception('Erreur lors de la récupération des récapitulatifs');
    }
  }
} 

File: Backend/src/main/java/com/example/demo/services/ReservationService.java
package com.example.demo.services;

import com.example.demo.entities.Reservation;

import java.security.Principal;
import java.util.List;
import java.util.Map;

public interface ReservationService {
    Reservation createReservation(Principal principal, Reservation reservation);
    
    // Récupérer toutes les réservations d'un utilisateur
    List<Reservation> getUserReservations(Principal principal);
    
    // Récupérer les réservations par statut pour un utilisateur
    List<Reservation> getUserReservationsByStatus(Principal principal, String status);
    
    // Mettre à jour une réservation avec un camion
    Reservation updateReservationWithCamion(Long reservationId, Long camionId, Principal principal,Boolean isIgnore);
    

    // Récupérer les réservations disponibles pour les transporteurs
    List<Reservation> getAvailableReservations();
    
    // Récupérer les réservations disponibles par statut pour les transporteurs
    List<Reservation> getAvailableReservationsByStatus(String status);
    
    // Récupérer toutes les réservations par statut (avec ou sans camion assigné)
    List<Reservation> getAllReservationsByStatus(String status);
    
    // Accepter une réservation (assigner le camion du transporteur)
    Reservation acceptReservation(Long reservationId, Principal principal);
    
    // Méthodes pour les réservations du transporteur
    List<Reservation> getMyReservations(Principal principal);
    List<Reservation> getMyReservationsByStatus(Principal principal, String status);
    Reservation updateReservationStatus(Long reservationId, String newStatus, Principal principal);
    
    // Nouvelle méthode pour retourner les réservations avec les informations du chargeur
    List<Map<String, Object>> getMyReservationsWithChargeurInfo(Principal principal, String status);
    
    // Méthode pour les statistiques du transporteur
    Map<String, Object> getMyStatistics(Principal principal);
    
    // Méthode de debug pour vérifier les données
    Map<String, Object> debugMyReservations(Principal principal);
    // Récupérer les récapitulatifs de réservations
    List<Map<String, Object>> getReservationRecapitulatif(Principal principal);
    

    
    // Récupérer une réservation par ID
    Reservation getReservationById(Long reservationId);

} 